lib/SILOptimizer/Transforms/DeadObjectElimination.cpp

```diff
@@ -829,21 +844,44 @@ bool DeadObjectElimination::processAllocRef(AllocRefInstBase *ARI) {
     return false;
   }
 
-  // Find the instruction which releases the object's tail elements.
-  SILInstruction *releaseOfTailElems = nullptr;
-  for (SILInstruction *user : UsersToRemove) {
-    if (isDestroyArray(user) ||
-       (destructorEffects == DestructorEffects::DestroysTailElems &&
-        isa<RefCountingInst>(user) && user->mayRelease())) {
-      // Bail if we find multiple such instructions.
-      if (releaseOfTailElems)
+  if (!ARI->getFunction()->hasOwnership()) {
+    // Find the instruction which releases the object's tail elements.
+    SILInstruction *releaseOfTailElems = nullptr;
+    for (SILInstruction *user : UsersToRemove) {
+      if (isDestroyArray(user) ||
+          (destructorEffects == DestructorEffects::DestroysTailElems &&
+           isa<RefCountingInst>(user) && user->mayRelease())) {
+        // Bail if we find multiple such instructions.
+        if (releaseOfTailElems)
+          return false;
+        releaseOfTailElems = user;
+      }
+    }
+    if (releaseOfTailElems) {
+      if (!insertCompensatingReleases(releaseOfTailElems, UsersToRemove)) {
         return false;
-      releaseOfTailElems = user;
+      }
     }
   }
-  if (releaseOfTailElems) {
-    if (!insertCompensatingReleases(releaseOfTailElems, UsersToRemove)) {
-      return false;
+
+  if (ARI->getFunction()->hasOwnership()) {
+    for (auto *user : UsersToRemove) {
+      auto *store = dyn_cast<StoreInst>(user);
+      if (!store ||
+          store->getOwnershipQualifier() == StoreOwnershipQualifier::Trivial)
+        continue;
+      if (findPointerEscape(store->getSrc())) {
+        return false;
+      }
+    }
+    for (auto *user : UsersToRemove) {
+      auto *store = dyn_cast<StoreInst>(user);
+      if (!store ||
+          store->getOwnershipQualifier() == StoreOwnershipQualifier::Trivial) {
+        continue;
+      }
+      SILBuilderWithScope(store).createDestroyValue(store->getLoc(),
+                                                    store->getSrc());
```

Comment:
When we destroy the element of the dead array when it is stored to the array (instead of when the dead array itself is destroyed), are we risking to illegally shrink the lifetime of the element? In case it's a lexical value.

Reply:
Thanks. I need a lexical check as well.